Content/Items/Haunted/Weapons.EchochainWhip.cs

```diff
@@ -0,0 +1,703 @@
+
+				Dust.NewDustPerfect(Projectile.Center + Main.rand.NextVector2Circular(5f, 5f), ModContent.DustType<Dusts.GlowFastDecelerate>(), Main.rand.NextVector2Circular(1f, 1f), 0, new Color(150, 255, 25), 0.65f);
+			}
+			else
+			{
+				Projectile.Center = Vector2.Lerp(Projectile.Center, tilePosition + new Vector2(0f, target.height > 25 ? -target.height : -25f), EaseBuilder.EaseQuarticIn.Ease((StabTimer - maxStabTimer * 0.25f) / (maxStabTimer * 0.75f)));
+				
+				if (target.knockBackResist > 0f)
+					target.Center = Projectile.Center - tilePosition.DirectionTo(target.Center) * 50f * (1f - progress);
+
+				if (progress >= 0.8f && !hitGround)
+				{
+					hitGround = true;
+					Terraria.Audio.SoundEngine.PlaySound(SoundID.DD2_MonkStaffGroundImpact, Projectile.Center);
+					CameraSystem.shake += 4;
+
+					target.SimpleStrikeNPC(Projectile.damage * 2, 0);
+
+					for (int i = 0; i < 20; i++)
+					{
+						Dust.NewDustPerfect(target.Center, ModContent.DustType<Dusts.GlowFastDecelerate>(), Main.rand.NextVector2Circular(5f, 5f), 0, new Color(150, 255, 25), 0.65f);
+					}
+
+					if (Helper.IsFleshy(target))
+					{
+						for (int i = 0; i < 15; i++)
+						{
+							Dust.NewDustPerfect(target.Center, ModContent.DustType<Dusts.GraveBlood>(), Main.rand.NextVector2Circular(5f, 5f), 0, default, 1.65f);
+
+							Dust.NewDustPerfect(target.Center, DustID.Blood, Main.rand.NextVector2Circular(15f, 15f), 100, default, 1.65f).noGravity = true;
+						}
+					}
+				}
+			}
+
+			if (Main.rand.NextBool(5))
+				Dust.NewDustPerfect(Projectile.Center + Main.rand.NextVector2Circular(5f, 5f), ModContent.DustType<Dusts.GlowFastDecelerate>(), Main.rand.NextVector2Circular(5f, 5f), 0, new Color(150, 255, 25), 0.5f);
+		}
+
+		public override void OnHitNPC(NPC target, NPC.HitInfo hit, int damageDone)
+		{
+			hasHit = true;
+			if (target.knockBackResist <= 0f || Main.projectile.Any(p => p.active && p != Projectile && p.type == Type && (p.ModProjectile as EchochainWhipAltProjectileChain).target == target))
+				Projectile.Kill();
+
+			CameraSystem.shake += 2;
+
+			for (int i = 0; i < 20; i++)
+			{
+				Dust.NewDustPerfect(target.Center, ModContent.DustType<Dusts.GlowFastDecelerate>(), tilePosition.DirectionTo(target.Center).RotatedByRandom(0.5f) * Main.rand.NextFloat(15f), 0, new Color(150, 255, 25), 0.65f);
+			}
+
+			if (Helper.IsFleshy(target))
+			{
+				for (int i = 0; i < 15; i++)
+				{
+					Dust.NewDustPerfect(target.Center, ModContent.DustType<Dusts.GraveBlood>(), tilePosition.DirectionTo(target.Center).RotatedByRandom(0.5f) * Main.rand.NextFloat(15f), 0, default, 1.65f);
+
+					Dust.NewDustPerfect(target.Center, DustID.Blood, tilePosition.DirectionTo(target.Center).RotatedByRandom(0.5f) * Main.rand.NextFloat(15f), 100, default, 1.65f).noGravity = true;
+				}
+			}
+
+
+			target.velocity += tilePosition.DirectionTo(target.Center) * 5f;
```

Comment:
on hit NPC and all the other onhits are not synced normally so if you do something like this with npc velocity / setting vars you need to sync it


you can use something like 
	Owner.TryGetModPlayer<StarlightPlayer>(out StarlightPlayer starlightPlayer);
	starlightPlayer.SetHitPacketStatus(shouldRunProjMethods: true);
to utilize our existing packet. this paradigm in particular will make it so all clients and the server run this onhit logic too. if that's undesirable you may want other methods of syncing (though looking at this particular use case I think its ideal)